### PR TITLE
X509 certificate implementation

### PR DESCRIPTION
**Pull request by cmanset**
First try to implement x509 certificates support in Leshan - work in progress

### CODE REVIEW

File: leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/SecureIntegrationTestHelper.java [hunk 1]
```diff
@@ -83,7 +95,14 @@ public SecureIntegrationTestHelper() {
...
             clientPublicKey = KeyFactory.getInstance("EC").generatePublic(publicKeySpec);
             clientPrivateKey = KeyFactory.getInstance("EC").generatePrivate(privateKeySpec);
-        } catch (InvalidKeySpecException | NoSuchAlgorithmException | InvalidParameterSpecException e) {
+
+            // Get certificates
+            // TODO handmade credentials -> automatically generated creds to delete after test
+            clientCAX509Cert = createCertificate("credentials/clientCA.crt");
+            clientX509Cert = createCertificate("credentials/client.crt");
+
+        } catch (InvalidKeySpecException | NoSuchAlgorithmException | InvalidParameterSpecException
```
Comment (jvermillard) [inline leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/SecureIntegrationTestHelper.java]: catch GeneralSecurityException ?

File: leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/SecureIntegrationTestHelper.java [hunk 2]
```diff
@@ -146,6 +180,41 @@ public void createRPKClient() {
...
+            InetSocketAddress clientAddress = new InetSocketAddress(InetAddress.getLoopbackAddress(), 0);
+            DtlsConnectorConfig.Builder config = new DtlsConnectorConfig.Builder(clientAddress);
+            config.setIdentity(clientPrivKey, clientX509CertChain, false);
+            config.setTrustStore(trustedCertificates);
+
+            CoapServer coapServer = new CoapServer();
+            coapServer.addEndpoint(new CoAPEndpoint(new DTLSConnector(config.build()), NetworkConfig.getStandard()));
+            client = new LeshanClient(clientAddress, server.getSecureAddress(), coapServer,
+                    new ArrayList<LwM2mObjectEnabler>(objects));
+        } catch (KeyStoreException | CertificateException | NoSuchAlgorithmException | IOException
```
Comment (jvermillard) [inline leshan-integration-tests/src/test/java/org/eclipse/leshan/integration/tests/SecureIntegrationTestHelper.java]: GeneralSecurityException ?

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/impl/SecurityRegistryImpl.java [hunk 3]
```diff
@@ -185,6 +194,15 @@ protected void saveToFile() {
     }
 
     @Override
+    public void setTrustedCertificates(Certificate[] trustedCertificates) {
+        if (trustedCertificates != null && trustedCertificates.length > 0) {
+            this.trustedCertificates = trustedCertificates;
+        } else {
+            LOG.debug("Tried to set trusted certificates to null");
```
Comment (jvermillard) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/impl/SecurityRegistryImpl.java]: nasty silent error :) use Validate.notNull?

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/security/SecurityInfo.java [hunk 4]
```diff
@@ -65,7 +71,15 @@ public static SecurityInfo newPreSharedKeyInfo(String endpoint, String identity,
...
     public static SecurityInfo newRawPublicKeyInfo(String endpoint, PublicKey rawPublicKey) {
         Validate.notNull(rawPublicKey);
-        return new SecurityInfo(endpoint, null, null, rawPublicKey);
+        return new SecurityInfo(endpoint, null, null, rawPublicKey, null);
+    }
+
+    /**
+     * Construct a {@link SecurityInfo} when using DTLS with an X509 Certificate.
+     */
+    public static SecurityInfo newX509CertInfo(String endpoint, X509Certificate[] X509CertChain) {
```
Comment (msangoi) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/security/SecurityInfo.java]: Is it needed to provide a certificate chain?
I think that we should call something like securityInfo.isX509() in the RegistrationHandler.
Maybe we could have an enum in the SecurityInfo to get the type of security?

**Comment by msangoi** [comment]
The last step would be to add the X509 type in the drop-down list of leshan-standalone UI :)

**Comment by sbernard31** [comment]
This is finally integrated in master (commit 4313f7ccde079618469e3532beca7279943e1809)

**Pull request closed without merging**